kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala

```diff
@@ -118,6 +118,14 @@ class KyuubiSessionImpl(
     engineLastAlive = System.currentTimeMillis()
   }
 
+  def reLaunchEngine(): Unit = handleSessionException {
+    _client.closeSession()
+    engineLaunched = false
+    launchEngineOp = sessionManager.operationManager
+      .newLaunchEngineOperation(this, shouldRunAsync = false)
+    runOperation(launchEngineOp)
```

Comment:
If in non-CONNECTION mode, this might still be routed to the same engine since we didn't kill it.

Reply:
In non-CONNECTION mode, how about current session not kill but start a new engine？